pyfr/backends/base/types.py

```diff
@@ -16,33 +17,47 @@ def __init__(self, backend, dtype, ioshape, initval, extent, aliases,
         self.dtype = dtype
         self.itemsize = np.dtype(dtype).itemsize
 
-        # Alignment requirement for the leading dimension
-        ldmod = backend.alignb // self.itemsize if 'align' in tags else 1
-
         # Our shape and dimensionality
         shape, ndim = list(ioshape), len(ioshape)
 
+        # AoSoA Size if present
+        aosoasz = backend.aosoasz if backend.cacheblks else 1
+
         if ndim == 2:
             nrow, ncol = shape
             aosoashape = shape
+            self.nbcol = backend.soasz*aosoasz
         else:
             nvar, narr, k = shape[-2], shape[-1], backend.soasz
-            nparr = narr - narr % -k
+            nparr = narr - narr % -(k*aosoasz)
 
             nrow = shape[0] if ndim == 3 else shape[0]*shape[1]
             ncol = nvar*nparr
             aosoashape = shape[:-2] + [nparr // k, nvar, k]
+            self.nbcol = nvar*backend.soasz*aosoasz
 
         # Assign
         self.nrow, self.ncol = int(nrow), int(ncol)
 
         self.datashape = aosoashape
         self.ioshape = ioshape
 
-        self.leaddim = self.ncol - (self.ncol % -ldmod)
+        # Alignment requirement for the leading dimension
+        if backend.cacheblks:
+            ldmod = backend.soasz*aosoasz if 'opmat' not in tags else 1
+            self.leaddim = self.nbcol
+            if 'opmat' in self.tags:
+                self.leaddim = self.nbcol = self.ncol
+        else:
+            ldmod = backend.soasz if 'align' in tags else 1
+            self.leaddim = self.ncol - (self.ncol % -ldmod)
+
+        self.blocksz = self.nrow*self.nbcol
+        self.nblock = math.ceil(self.ncol/self.nbcol)
```

Comment:
Don't round up like this.  Use modular arithmetic like we do elsewhere in the function.